### PR TITLE
fix(browser): retry CDP screenshots and log failures with context

### PR DESCRIPTION
**Pull request by leo-notte**
CDP screenshot previously had one shot before silently falling back to Playwright on debug-level logs, so hangs and transient failures were invisible in production. Add a bounded retry loop (3 attempts, 10s each) with per-stage timing (attach/send/detach), full exception info on failure, and slow-detach warnings to surface CDP-session contention.

## Summary by CodeRabbit

* **Improvements**
  * More reliable screenshot captures with bounded timeouts and capped retry attempts
  * Distinguishes timeout failures from other errors, waits between retries, and logs per-attempt timings
  * Stronger fallback behavior to alternate capture method when primary capture fails
  * Clearer, more informative warnings and diagnostics to reduce hangs and aid troubleshooting

<h3>Greptile Summary</h3>

This PR replaces the single-shot CDP screenshot with a bounded retry loop (up to 3 attempts at 10 s each), adds per-stage timing diagnostics (attach/send/detach), and shields the session detach from `asyncio.wait_for` cancellation to prevent CDP session leaks that the prior implementation was vulnerable to. The previously flagged P1 (dangling sessions on timeout) is resolved by `asyncio.shield(asyncio.wait_for(detach, 2.0))`; the no-tests concern was explicitly deferred by the team.

<h3>Confidence Score: 5/5</h3>

Safe to merge — the prior P1 (CDP session leak on timeout) is addressed and no new P0/P1 issues are introduced.

All remaining findings are P2. The previously raised CDP session-leak concern is resolved with asyncio.shield. There is one minor P2: if the shielded background detach task itself times out, nobody awaits the resulting TimeoutError, which can produce a Python 'Task exception was never retrieved' warning in pathological cases. This does not affect correctness or crash anything.

No files require special attention.

<h3>Important Files Changed</h3>

| Filename | Overview |
|----------|----------|
| packages/notte-browser/src/notte_browser/window.py | Adds CDP screenshot retry loop (3 attempts, 10s each), per-stage timing logs, and shielded detach to prevent session leaks on timeout. Previously flagged P1 (session leak on cancellation) is addressed. |

</details>

<a href="https://app.greptile.com/ide/claude-code?prompt=Fix%20the%20following%201%20code%20review%20issues.%20Work%20through%20them%20one%20at%20a%20time%2C%20proposing%20concise%20fixes.%0A%0A---%0A%0A%23%23%23%20Issue%201%20of%201%0Apackages%2Fnotte-browser%2Fsrc%2Fnotte_browser%2Fwindow.py%3A349-352%0A**Unhandled%20exception%20in%20shielded%20background%20task**%0A%0AWhen%20the%20outer%20%60wait_for%60%20cancels%20%60_run%60%2C%20%60asyncio.shield%60%20raises%20%60CancelledError%60%20here%20%28caught%20by%20%60pass%60%29%2C%20but%20the%20inner%20%60wait_for%28detach%2C%202.0%29%60%20continues%20running%20as%20an%20independent%20background%20task.%20If%20%60detach%60%20then%20stalls%20past%202%20s%2C%20that%20inner%20%60wait_for%60%20raises%20%60TimeoutError%60%20inside%20the%20background%20task%20%E2%80%94%20and%20since%20nobody%20ever%20awaits%20the%20shielded%20task%20after%20%60CancelledError%60%20is%20caught%2C%20Python%20will%20emit%20a%20%22Task%20exception%20was%20never%20retrieved%22%20warning.%20To%20suppress%20it%2C%20you%20can%20attach%20a%20no-op%20callback%20to%20the%20shielded%20task%3A%0A%0A%60%60%60python%0Ainner%20%3D%20asyncio.ensure_future%28asyncio.wait_for%28cdp_session.detach%28%29%2C%20timeout%3D2.0%29%29%0Ainner.add_done_callback%28lambda%20t%3A%20t.exception%28%29%20if%20not%20t.cancelled%28%29%20else%20None%29%0Atry%3A%0A%20%20%20%20await%20asyncio.shield%28inner%29%0Aexcept%20asyncio.CancelledError%3A%0A%20%20%20%20pass%0Aexcept%20Exception%20as%20e%3A%0A%20%20%20%20logger.warning%28f%22CDP%20session%20detach%20for%20%7Bself.page.url%7D%20failed%3A%20%7Btype%28e%29.__name__%7D%3A%20%7Be%7D%22%29%0A%60%60%60%0A%0A&repo=nottelabs%2Fnotte&pr=782&platform=github"><picture><source media="(prefers-color-scheme: dark)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixAllInClaudeDark.svg?v=2"><source media="(prefers-color-scheme: light)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixAllInClaude.svg?v=2">Fix All in Claude Code</picture></a>

<details><summary>Prompt To Fix All With AI</summary>

`````markdown
This is a comment left during a code review.
Path: packages/notte-browser/src/notte_browser/window.py
Line: 349-352

Comment:
**Unhandled exception in shielded background task**

When the outer `wait_for` cancels `_run`, `asyncio.shield` raises `CancelledError` here (caught by `pass`), but the inner `wait_for(detach, 2.0)` continues running as an independent background task. If `detach` then stalls past 2 s, that inner `wait_for` raises `TimeoutError` inside the background task — and since nobody ever awaits the shielded task after `CancelledError` is caught, Python will emit a "Task exception was never retrieved" warning. To suppress it, you can attach a no-op callback to the shielded task:

```python
inner = asyncio.ensure_future(asyncio.wait_for(cdp_session.detach(), timeout=2.0))
inner.add_done_callback(lambda t: t.exception() if not t.cancelled() else None)
try:
    await asyncio.shield(inner)
except asyncio.CancelledError:
    pass
except Exception as e:
    logger.warning(f"CDP session detach for {self.page.url} failed: {type(e).__name__}: {e}")
```

How can I resolve this? If you propose a fix, please make it concise.
`````

</details>

<sub>Reviews (2): Last reviewed commit: ["fix(browser): shield CDP detach from wai..."](https://github.com/nottelabs/notte/commit/ad0a1f49585df7334542fd30d86a179807248bce) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=29361182)</sub>

---

> [!NOTE]
> Replaces the single-shot CDP screenshot with a bounded retry loop (3 attempts, 10s each), adds per-stage timing diagnostics (attach/send/detach), shields CDP session detach from `asyncio.wait_for` cancellation to prevent session leaks, and propagates a `_skip_cdp` flag through recursion to prevent compounding retry budgets on `PlaywrightTimeoutError`.
> 
> <sup>Written by [Mendral](https://mendral.com) for commit 4d3dcf76cd89a0f7e055fbc3fa88b77e54e4d69f.</sup>

### CODE REVIEW

File: packages/notte-browser/src/notte_browser/window.py [hunk 1]
```diff
@@ -315,18 +315,43 @@ async def snapshot_metadata(self) -> SnapshotMetadata:
...
+                await cdp_session.detach()
+                detach_ms = (time.monotonic() - t_detach) * 1000
+                # Detach is normally <5ms — log anything slower so we can spot contention
+                # between concurrent CDP sessions sharing the same page.
+                if detach_ms > 500:
+                    logger.warning(
+                        f"CDP session detach for {self.page.url} took {detach_ms:.0f}ms (attach={attach_ms:.0f}ms)"
+                    )
+
+        return await asyncio.wait_for(_run(), timeout=self.CDP_SCREENSHOT_TIMEOUT_S)
```
[the remark this answers is not in the document]
Comment (leo-notte) [reply packages/notte-browser/src/notte_browser/window.py:377]: Skipping tests for this PR. `notte-browser` has no existing test harness for `BrowserWindow` or the screenshot path, and mocking the full CDP surface (session attach/send/detach plus the page) for a diagnostic-logging + retry-count change would be more glue than signal. The load-bearing behavior change is the shielded detach from the P1 thread, which is now in.

**Comment by leo-notte** [comment]
@greptileai review

File: packages/notte-browser/src/notte_browser/window.py [hunk 2]
```diff
@@ -315,18 +315,48 @@ async def snapshot_metadata(self) -> SnapshotMetadata:
...
+            finally:
+                t_detach = time.monotonic()
+                # Shield detach from outer wait_for cancellation — otherwise a timeout here
+                # would skip detach and leak the CDP session, which is the exact contention
+                # this path is trying to surface. Inner wait_for bounds detach itself so a
+                # hung detach can't live forever in the background task set.
+                try:
+                    await asyncio.shield(asyncio.wait_for(cdp_session.detach(), timeout=2.0))
+                except Exception as e:
+                    logger.warning(f"CDP session detach for {self.page.url} failed: {type(e).__name__}: {e}")
```
[the remark this answers is not in the document]
Comment (leo-notte) [reply packages/notte-browser/src/notte_browser/window.py]: Addressed in 0cf83625. The shielded detach task is now explicitly created (via `asyncio.ensure_future`) with a done callback attached, so its outcome (success, inner 2s timeout, or failure) is logged even when the outer `wait_for` is cancelled and leaves it running unobserved. The outer `except Exception` already does not swallow `CancelledError` (it inherits from `BaseException` in Py 3.8+), so cancellation propagates cleanly.

[the remark this answers is not in the document]
Comment (leo-notte) [reply packages/notte-browser/src/notte_browser/window.py]: Addressed in 0cf83625. The shielded detach task is now explicitly created (via `asyncio.ensure_future`) with a done callback attached, so its outcome (success, inner 2s timeout, or failure) is logged even when the outer `wait_for` is cancelled and leaves it running unobserved. The outer `except Exception` already does not swallow `CancelledError` (it inherits from `BaseException` in Py 3.8+), so cancellation propagates cleanly.

**Comment by leo-notte** [comment]
@greptileai review

File: packages/notte-browser/src/notte_browser/window.py [hunk 3]
```diff
@@ -335,30 +383,64 @@ async def screenshot(self, retries: int = config.empty_page_max_retry) -> bytes:
...
             raise last_error  # type: ignore[misc]
 
         except PlaywrightTimeoutError:
-            if config.verbose:
-                logger.debug(f"Timeout while taking screenshot for {self.page.url}. Retrying...")
+            logger.warning(
+                f"Playwright screenshot timeout for {self.page.url}, retrying (remaining outer retries: {retries - 1})"
+            )
             await self.short_wait()
             return await self.screenshot(retries=retries - 1)
```
[the remark this answers is not in the document]
Comment (leo-notte) [reply packages/notte-browser/src/notte_browser/window.py]: Addressed in 4d3dcf76. Added a private `_skip_cdp` kwarg to `screenshot()`, set `cdp_exhausted = True` after the CDP loop gives up, and pass it to the recursive call on `PlaywrightTimeoutError`. With the default `empty_page_max_retry=5` this prevents up to 5 × 30s of compounded CDP retries (~150s worst case) before Playwright gets a clean chance.

File: packages/notte-browser/src/notte_browser/window.py [hunk 4]
```diff
@@ -315,18 +315,66 @@ async def snapshot_metadata(self) -> SnapshotMetadata:
...
+                    elif elapsed_ms > 500:
+                        logger.warning(
+                            f"CDP session detach for {url} took {elapsed_ms:.0f}ms (attach={attach_ms:.0f}ms)"
+                        )
+
+                detach_task.add_done_callback(_log_detach)
+                try:
+                    await asyncio.shield(detach_task)
+                except Exception:
+                    pass  # _log_detach handles logging via the done callback
```
[the remark this answers is not in the document]
Comment (leo-notte) [reply packages/notte-browser/src/notte_browser/window.py:375]: Not applying this one — the project's ruff config (pyproject.toml [tool.ruff.lint]) only `extend-select = ["I"]` (isort). S110 and BLE001 are not enabled here, and there are already 74 BLE001 / 9 S110 occurrences across `packages/` (including lines 398 and 432 of this same file). Narrowing just these two lines would create local inconsistency without any check enforcing the broader rule, and the `_log_detach` done callback already logs the outcome (cancellation, exception type+msg, or slow completion) — the `except Exception: pass` exists precisely so we don't double-log.

**Comment by leo-notte** [comment]
@greptileai review
